packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1883,19 +1883,26 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
     newRoute.install(_currentOverlayEntry);
     _history[index] = newRoute;
     newRoute.didPush().whenCompleteOrCancel(() {
-      // The old route's exit is not animated. We're assuming that the
-      // new route completely obscures the old one.
       if (mounted) {
+        // The old route's exit is not animated. We're assuming that the
+        // new route completely obscures the old one.
         oldRoute
           ..didComplete(result ?? oldRoute.currentResult)
           ..dispose();
+        final int newIndex = _history.indexOf(newRoute);
+        if (newIndex > 0) {
+          final Route<dynamic> precedingRoute = _history[newIndex - 1];
+          precedingRoute.didChangeNext(newRoute);
+          newRoute.didChangePrevious(precedingRoute);
+        }
       }
     });
     newRoute.didChangeNext(null);
+    oldRoute.didChangeNext(newRoute);
     if (index > 0) {
       _history[index - 1].didChangeNext(newRoute);
-      newRoute.didChangePrevious(_history[index - 1]);
```

Comment:
Why do we no longer need to notify previous?

Reply:
Fair point, it could still have a previous notification - I'll add that back in, though the previous will now be `oldRoute` (the removing route), which is changed to the preceding route on completion of the push.

Reply:
Let's revert this?

Reply:
In theory, we should call didChangePrevious  with oldRoute first and then call with _history[index - 1] after the transition. However, it is not possible with current navigator, and that is part of the reason why we are refactoring the navigator. I am more leaning toward keeping the original behavior and wait for the refactor to clean up to avoid introducing unexpected issue

Reply:
Why is that not possible? I've pushed a commit doing exactly that. There clearly is not yet a consistent view on when to notify next/previous. For now, I have taken the view that we notify next/previous before a transition. And notify previous when we have removed items (after the transition) - leaving the newly pushed route in a state with correct next/previous. It may be that the new route has the same previous set twice, but surely the effect of that notification should be idempotent. As stated before, any following transition will re-notify as required anyway. This is no worse than the existing situation, and is not something we should keep changing as part of this PR.
I appreciate that you wish to focus on your own refactor, but I am working on a commercial project that requires this change and cannot afford to abandon this PR for one with a much less predictable time frame.

Reply:
Is there another issue that depends on this change (line 1899 didChangePrevious )? I thought the original issue about the secondary animation can be solved by line 1900 didChangeNext alone. I do agree the change on line 1900 is necessary, and this is an actual bug that we didn't notify didChangeNext on the old route before. However, I think we should add the ```newRoute.didChangePrevious(_history[index - 1]);``` back while removing line 1899 if there is no good reason to change it.

Reply:
Agreed - have pushed a change to revert this